src/coreComponents/common/FourMutexesLock.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 TotalEnergies
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+#ifndef FOURMUTEXESLOCK_HPP
+#define FOURMUTEXESLOCK_HPP
+
+#include <mutex>
+
+namespace geosx
+{
+
+/**
+ * @brief Class to handle locks using 4 mutexes
+ */
+class FourMutexesLock
```

Comment:
This and `TwoMutexesLock` could be unified in a variadic class for any number of mutexes (essentially a c++14 replacement for [scoped_lock](https://en.cppreference.com/w/cpp/thread/scoped_lock)). Without deduction guides the usage will be slightly less elegant, but still better than having duplicate classes.

Reply:
Indeed I supposed I could do something on this, just did not figured out how to do it easily.

Reply:
I guess something like that should do the trick: https://blazingcpp.wordpress.com/2020/11/21/stdtuple-and-parameter-packs/

Reply:
It's a good article, but it uses C++17. I made a little prototype of what I mean in C++14, feel free to reuse as much as you'd like: https://godbolt.org/z/n4KdnWWrv

Reply:
Thanks, I ended with a similar code (thanks to chatGPT also ^^) but I understood that std::apply was C++17 also. It seems not so that perfect.

Reply:
I replaces one of my TwoMutexLock with this MultiMutexesLock() and get a deadlock for now:
```
/*
 * ------------------------------------------------------------------------------------------------------------
 * SPDX-License-Identifier: LGPL-2.1-only
 *
 * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
 * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
 * Copyright (c) 2018-2020 TotalEnergies
 * Copyright (c) 2019-     GEOSX Contributors
 * All rights reserved
 *
 * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
 * ------------------------------------------------------------------------------------------------------------
 */
#ifndef MULTIMUTEXESLOCK_HPP
#define MULTIMUTEXESLOCK_HPP

#include <mutex>
#include <tuple>

#include "codingUtilities/Utilities.hpp"

namespace geosx
{
// apply implementation taken from https://en.cppreference.com/w/cpp/utility/apply
namespace internal
{

template < class F, class Tuple, std::size_t... I >
constexpr decltype( auto ) apply( F&& f, Tuple&& t, std::index_sequence<I...> )
{
    return std::forward<F>( f )( std::get<I>(std::forward<Tuple>(t))...);
}

}

template < class F, class Tuple >
constexpr decltype(auto) apply( F&& f, Tuple&& t )
{
    return internal::apply( std::forward< F >( f ), std::forward< Tuple >( t ),
                            std::make_index_sequence< std::tuple_size< std::remove_reference_t<Tuple> >::value >{} );
}

/**
 * @brief Class to handle locks using 2 mutexes
 */
template < typename... Mutexes >
class MultiMutexesLock
{
public:
  /**
   * @brief Construct a multi mutexes lock and lock the mutexes.
   * @param mutexes The mutexes associated with the lock.
   */
    MultiMutexesLock( Mutexes&... mutexes )
      : m_islocked(false), m_mutexes(std::make_tuple(std::ref(mutexes)...))
    {
      lock();
    }

    void lock()
    {
      if( m_islocked ) return;
      apply( []( auto && ... mutexes, auto ){ std::lock( mutexes... ); }, m_mutexes );
      m_islocked = true;
    }

    void unlock()
    {
      if( !m_islocked ) return;
      forEachArgInTuple( m_mutexes, []( auto & mutex ){ mutex.unlock(); } );
      m_islocked = false;
    }

private:
  /// Indicate if the mutexes are owned by the lock
  bool m_islocked;
  /// Array of references to the mutexes
  std::tuple<Mutexes&...> m_mutexes;
};
}
#endif // MULTIMUTEXESLOCK_HPP
```

Reply:
The code you posted is missing a destructor `~MultiMutexesLock()` which could explain the deadlock. Can you check if that is the case? And while we're at it, I would also add a deleted copy constructor to prevent accidental misuse. 

Another question I had was whether you actually need the publicly accessible `lock()` and `unlock()` functions. In my example code I made them private on purpose to emulate the purely scope-based locking control that `std::scoped_lock` implements (it also makes the code simpler, since you no longer need the `m_islocked` flag). It's up to you though.

Reply:
Indeed I never use them directly. I'll make them private, and yes I was never calling the unlock() this is why I got the deadlock...

Reply:
I need the publicity for waiting on condition variables.

Reply:
Seems I get a race condition, need investigate that.

Reply:
I didn't copied your code correctly it seems (and running on my home directory didn't help...) No it seems OK, I removed the std::make_tuple(std::ref(mutexes)) and used your make_multilock(), tests passed fine.